### PR TITLE
[WIP] Dialog routing implementation

### LINKED ISSUE #1: Add random access navigation
Random Access Navigation is way for us to introduce a dozen ways to propose new sessions for the weekly wednesday sessions without forcing us to use very long dialog trees. 

See this article for more information: https://medium.com/assist/theres-a-dozen-ways-to-order-a-coffee-why-do-dumb-bots-only-allow-one-27230542636d

### PR DESCRIPTION
**Pull request by wmeints**
This PR contains all the code needed to implement the random access navigation.

I took the MS bot framework layout for dialogs as an inspiration. I didn't implement all their features, because there's quite a few things going on. Instead I kept it simple for now. You can setup a dialog and reply to messages.

This implements a workable solution #1

- [x] Implement child dialogs
- [x] Add a way to store information in context of a chat
- [ ] Implement propose session dialog
- [x] Use futures to start the slack connector
- [x] Add better javadoc to explain the dialog system
- [x] Replace custom factory method with `Supplier<T>`
- [x] Use lombok as much as we can
- [x] Add tests for different parts of the dialog system
- [x] Fix generating the `fromJson` methods in generated `Converter` classes
- [x] Test dialog

### CODE REVIEW

File: src/main/java/nl/infosupport/iskabot/chat/Dialog.java [hunk 1]
```diff
@@ -0,0 +1,12 @@
+package nl.infosupport.iskabot.chat;
+
+/**
+ * Implement {@llink Dialog} to create your own dialogs
```
Comment (mthmulders) [inline src/main/java/nl/infosupport/iskabot/chat/Dialog.java]: The correct Javadoc syntax is `{@link Dialog}`.

File: src/main/java/nl/infosupport/iskabot/chat/DialogContextImpl.java [hunk 2]
```diff
@@ -0,0 +1,82 @@
...
+    private final Handler<DialogContext> finalizeHandler;
+
+    /**
+     * Initializes {@link DialogContextImpl}
+     *
+     * @param sessionId       Session ID
+     * @param messageSink     Message sink to use for posting messages to the user
+     * @param finalizeHandler Finalizer handler for the dialog context
+     */
+    public DialogContextImpl(String sessionId, UserMessageSink messageSink, Handler<DialogContext> finalizeHandler) {
```
Comment (mthmulders) [inline src/main/java/nl/infosupport/iskabot/chat/DialogContextImpl.java]: Could this be replaced by Lombok's `@AllArgsConstructor` (and moving the assignment of `this.handlers` to the place where it is declared)?

File: src/main/java/nl/infosupport/iskabot/chat/DialogContextImpl.java [hunk 3]
```diff
@@ -0,0 +1,82 @@
...
+    public void wait(DialogHandler handler) {
+        handlers.push(handler);
+    }
+
+    /**
+     * Ends the conversation
+     */
+    @Override
+    public void end() {
+        if (handlers.empty()) {
```
Comment (mthmulders) [inline src/main/java/nl/infosupport/iskabot/chat/DialogContextImpl.java]: What happens if there are handlers? It seems the conversation is not ended in that case, but there is no logging of that either.

File: src/main/java/nl/infosupport/iskabot/chat/DialogContextImpl.java [hunk 4]
```diff
@@ -0,0 +1,82 @@
...
+
+    /**
+     * Handle incoming message
+     *
+     * @param activity Activity to handle
+     */
+    public void handle(ChatActivity activity) {
+        DialogHandler handler = handlers.pop();
+
+        if (handler != null) {
```
Comment (mthmulders) [inline src/main/java/nl/infosupport/iskabot/chat/DialogContextImpl.java]: What if there are no handlers?

File: src/main/java/nl/infosupport/iskabot/chat/UserMessageSink.java [hunk 5]
```diff
@@ -0,0 +1,12 @@
+package nl.infosupport.iskabot.chat;
+
+/**
+ * Implement this {@link UserMessageSink} to post messages back to the user
```
Comment (mthmulders) [inline src/main/java/nl/infosupport/iskabot/chat/UserMessageSink.java]: Do I understand correctly that this interface is supposed to be implemented by protocols that allow user communication (e.g. an API response, a Slack conversations)?

File: src/main/java/nl/infosupport/iskabot/slack/SlackBotVerticle.java [hunk 6]
```diff
@@ -14,49 +16,44 @@
...
-                this.slackRtmClient = slack.rtm(configuration.getSlackApiToken());
-                this.slackRtmClient.addMessageHandler(this::handleIncomingSlackMessage);
-                this.slackRtmClient.connect();
-                logger.info("Connected to Slack...");
-                start.complete();
-            } catch (IOException | DeploymentException e) {
-                start.fail(e);
-            }
-        }, (ar) -> { /* ignore */ });
+    public void start() throws Exception {
```
Comment (mthmulders) [inline src/main/java/nl/infosupport/iskabot/slack/SlackBotVerticle.java]: The reason for having this executed inside `vertx.executeBlocking` is that you shouldn't block the event loop for too long. Connecting to Slack first and then connecting to the WebSocket might take some time and is definitely blocking on network I/O. Why did you remove the invocation of `vertx.executeBlocking`?

File: src/main/java/nl/infosupport/iskabot/slack/SlackBotVerticle.java [hunk 7]
```diff
@@ -14,49 +16,44 @@
...
+     */
     private void handleIncomingSlackMessage(final String rawMessage) {
-        final SlackMessage message;
-        try {
-            message = jsonParser.fromJson(rawMessage, SlackMessage.class);
-        } catch (final JsonSyntaxException jse) {
-            logger.error("Could not parse message from Slack. Original message: {}", rawMessage, jse);
-            return;
-        }
+        final SlackMessage message = Json.decodeValue(rawMessage, SlackMessage.class);
```
Comment (mthmulders) [inline src/main/java/nl/infosupport/iskabot/slack/SlackBotVerticle.java]: Ah, more elegant! Thanks!

File: src/main/java/nl/infosupport/iskabot/slack/SlackBotVerticle.java [hunk 8]
```diff
@@ -65,75 +62,43 @@ private void handleIncomingSlackMessage(final String rawMessage) {
...
-    private Future<IntentDetectionResult> detectIntent(final String text) {
-        final Future<IntentDetectionResult> result = Future.future();
+    /**
+     * Retrieves an existing chat conversation or starts a new one if none exists.
+     *
+     * @param sessionId ID of the session (Usually the username for the user)
+     * @param initFunc  The factory method that produces the root dialog
+     * @return Returns the dialog context implementation to use
+     */
+    private DialogContextImpl getOrCreateContext(String sessionId, DialogFactoryFunction initFunc) {
```
Comment (mthmulders) [inline src/main/java/nl/infosupport/iskabot/slack/SlackBotVerticle.java]: Why have a separate interface `DialogFactoryFunction` when we have `java.util.function.Supplier<DialogFactory>`?

Comment (wmeints) [inline src/main/java/nl/infosupport/iskabot/slack/SlackBotVerticle.java]: Good point. I need to fix that. Didn't know about the supplier interface.

File: src/main/java/nl/infosupport/iskabot/slack/SlackBotVerticle.java (hunk 6 above)
Comment (wmeints) [inline src/main/java/nl/infosupport/iskabot/slack/SlackBotVerticle.java]: For me it was really short, but I can imagine that we need to make this more robust. I actually found out that the Future implementation of vert.x is even more elegant. I will add that instead.

File: src/main/java/nl/infosupport/iskabot/chat/UserMessageSink.java (hunk 5 above)
Comment (wmeints) [inline src/main/java/nl/infosupport/iskabot/chat/UserMessageSink.java]: Yup, that's the idea. I am working on some tests and this allows me to disconnect the dialog context from the rest of the app. Not sure if it remains this way. Not really happy with it yet.

File: src/main/java/nl/infosupport/iskabot/chat/DialogContextImpl.java (hunk 3 above)
Comment (wmeints) [inline src/main/java/nl/infosupport/iskabot/chat/DialogContextImpl.java]: This is actually part of a construct that I am still working on. The idea is that you can have child dialogs, that push additional handlers on the stack. When you and a child dialog it should leave that dialog (not pushing a new handler for that dialog on the stack), but not end the conversation as a whole.

File: src/main/java/nl/infosupport/iskabot/chat/UserMessageSink.java (hunk 5 above)
Comment (wmeints) [inline src/main/java/nl/infosupport/iskabot/chat/UserMessageSink.java]: I'm going to keep the construct as it is. It's great for decoupling the `SlackbotVerticle` from the `DialogContextImpl`. The tests were way easier to write with this in place :-)

**Comment by wmeints** [comment]
Looking good. I think this one's done with the new tests in place.
